### DISCUSSION TITLE
LINQ Any translated differently from v2

### DISCUSSION THREAD

**Discussion by hkarask** [linq2db/linq2db#3429] [Q&A]
Hi all,

I'm trying to bump the `linq2sql` version in an older codebase, from `1.10.2` to `2.0.0` and noticed that `LINQ` `Any` gets translated a bit differently:

Given a contrived LINQ statement:

```csharp
var query =
    from task in con.GetTable<Task>()
    from party in con.GetTable<Party>()
        .Where(p => task.AdminPartyId == p.Id ||
                    con.GetTable<PartyAccess>().Any(pa => pa.PartyId == p.Id && pa.Role == Roles.Admin)
        ).DefaultIfEmpty()
...
```

I'm seeing different SQL between versions 1 and 2

### v 1.10.2
```sql
SELECT
 	t.Description,
 	p.Name
 FROM
 	Task t
 		LEFT JOIN Party p ON t.AdminPartyId = p.Id OR EXISTS(
 			SELECT
 				*
 			FROM
 				PartyAccess pa
 			WHERE
 				pa.PartyId = p.Id AND pa.Role = 'Admin'
 		)
```
### v 2.0.0
```sql
SELECT
 	j.Description,
 	t2.Name
 FROM
 	Task t
 		LEFT JOIN (
 			SELECT
 				p.Name,
 				p.Id
 			FROM
 				Party p
 			WHERE
 				(EXISTS(
 					SELECT
 						*
 					FROM
 						PartyAccess pa
 					WHERE
 						pa.PartyId = p.Id AND pa.Role = 'Admin'
 				))
 		) t2 ON t.AdminPartyId = t2.Id
```

There's now a full sub-query instead of join a level filter. This is causing a full table scan for some reason (using MySql 5.6). Is there a way to revert back to the old behaviour?

## 

### Environment details
linq2db version: `1.10.2` & `2.0.0`
Database Server: `MySql 5.6`
Database Provider: `MySql.Data 8.0.25`
.NET Framework: `.NET 4.8`

Thanks!

**Comment by MaceWindu** [comment]
2.0 was released almost 4 years ago and a lot of stuff changed since than. I recomment to try more recent version, as I doubt anyone could advice you on v2 now.

**Reply by MaceWindu** [reply]
I've tried your query on current version and observe similar issues, so I've created #3432 issue to track it

**Reply by hannesAtPropertyMe** [reply]
Thanks for the quick reply @MaceWindu

**Comment by Usaga** [comment]
@hkarask you can workaround this:

```c#
            var query = con.GetTable<Task>()
                .LeftJoin(
                    con.GetTable<Party>(),
                    (t, p) => t.AdminPartyId == p.Id ||
                    con.GetTable<PartyAccess>().Any(pa => pa.PartyId == p.Id && pa.Role == Roles.Admin),
                    (t, p) => new
                    {
                        t.Description,
                        p.Name
                    })
                    .ToList();
```

This gives exactly what you want:

```sql
SELECT
	`t`.`Description`,
	`p`.`Name`
FROM
	`Task` `t`
		LEFT JOIN `Party` `p` ON (`t`.`AdminPartyId` = `p`.`Id` OR EXISTS(
			SELECT
				*
			FROM
				`PartyAccess` `pa`
			WHERE
				`pa`.`PartyId` = `p`.`Id` AND `pa`.`Role` = 'Admin'
		))
```

**Reply by hannesAtPropertyMe** [reply]
This is generating the query I need, thank you @Usaga.

However the codebase I'm working on has a `DbContext` class which has multiple getters exposing `IQueryables`, e.g

```csharp
class DbContext 
{
    public IQueryable<Task> Tasks =>_con.GetTable<Task>().Where(t => t.IsActive);

    public IQueryable <Party> Admins => con.GetTable<Party>()
        .Where(p => con.GetTable<PartyAccess>().Any(pa => pa.PartyId == p.Id && pa.Role == Roles.Admin));

    public IQueryable <Project> Projects => con.GetTable<Project>().Where(t => t.IsActive); 
}
```

And in the consumer class we combine those together:

```csharp
class Consumer
{
    from task in context.Tasks
    from admin in context.Admins.Where(p => task.AdminPartyId == p.Id).DefaultIfEmpty()
    from project in context.Projects.Where(p => task.ProjectId == p.Id).DefaultIfEmpty()
    select new { task, admin, project }
}
```
Is there a way I could have similar composition, given that the `LeftJoin<TOuter, TInner, TResult>(outer, inner, predicate, resultSelector)` needs a different approach.
